src/Common/ICachePolicyUserQuota.h

```diff
@@ -0,0 +1,49 @@
+#pragma once
+
+#include <base/types.h>
+
+namespace DB
+{
+
+/// Per-user quotas for usage of shared caches, used by ICachePolicy.
+/// Currently allows to limit
+/// - the maximum amount of cache memory a user may consume
+/// - the maximum number of items a user can store in the cache
+/// Note that caches usually also have global limits which restrict these values at cache level. Per-user quotas have no effect if they
+/// exceed the global thresholds.
+class ICachePolicyUserQuota
+{
+public:
+    enum class Resource
+    {
+        Size,
+        Items
```

Comment:
Maybe also use the terms `bytes` and `entries`?

Actually, I do not see much sense in the separation of these two resources, because increase/decrease are always called for an entry that has both size in bytes and plain count (and any other resource as far as I can imagine), so you always have two calls to increase/decrease. Maybe it would be better to add `struct Resources { size_t bytes; size_t entries; }` instead of enum? It will also allow you to have just 2 maps in implementations instead of 2 * resource_types.

Reply:
Fair enough. The logic is quite overengineered ... I'll simplify in the way you suggested.